src/Servers/Kestrel/Core/src/Internal/PinnedBlockMemoryPoolFactory.cs

```diff
@@ -0,0 +1,46 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Collections.Concurrent;
+using System.Diagnostics.Metrics;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal;
+
+internal sealed class PinnedBlockMemoryPoolFactory : IMemoryPoolFactory<byte>, IHeartbeatHandler
+{
+    private readonly IMeterFactory _meterFactory;
+    private readonly TimeProvider _timeProvider;
+    private readonly ConcurrentDictionary<PinnedBlockMemoryPool, PinnedBlockMemoryPool> _pools = new();
```

Comment:
nit: Could ask @stephentoub to create a `ConcurrentHashTable<T>` 😉 or change the value to a recognised atomic type by `ConcurrentDictionary` e.g. `nuint`
```csharp
    private readonly ConcurrentDictionary<PinnedBlockMemoryPool, nuint> _pools = new();
```
Will save on GC write barriers; thought probably not too important

Reply:
Link to any documentation or discussion about this?

Reply:
Micro-optimization on the correct data structure not being available in runtime (as both key and value are always the same and value never read)